File: app/src/main/java/com/andb/apps/todo/objects/Tasks.java
package com.andb.apps.todo.objects;

import android.util.Log;

import com.andb.apps.todo.InboxFragment;
import com.andb.apps.todo.settings.SettingsActivity;
import com.andb.apps.todo.typeconverters.CheckedConverter;
import com.andb.apps.todo.typeconverters.ItemsConverter;
import com.andb.apps.todo.typeconverters.TagConverter;
import com.andb.apps.todo.utilities.Current;
import com.andb.apps.todo.utilities.ProjectsUtils;
import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

import org.joda.time.DateTime;

import java.util.ArrayList;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.ForeignKey;
import androidx.room.Ignore;
import androidx.room.PrimaryKey;
import androidx.room.TypeConverters;

import static com.andb.apps.todo.utilities.Values.TIME_SORT;

@Entity(foreignKeys = @ForeignKey(entity = BaseProject.class, parentColumns = "key", childColumns = "project_id", onDelete = ForeignKey.CASCADE))
public class Tasks implements Cloneable {
    @SerializedName("key")
    @Expose
    @PrimaryKey
    private int listKey;

    @SerializedName("list_name")
    @Expose
    @ColumnInfo(name = "list_name")
    private String listName;

    @SerializedName("list_items")
    @Expose
    @TypeConverters(ItemsConverter.class)
    @ColumnInfo(name = "list_items")
    private ArrayList<String> listItems;

    @SerializedName("list_items_checked")
    @Expose
    @TypeConverters(CheckedConverter.class)
    @ColumnInfo(name = "list_items_checked")
    private ArrayList<Boolean> listItemsChecked;

    @SerializedName("list_tags")
    @Expose
    @TypeConverters(TagConverter.class)
    @ColumnInfo(name = "list_tags")
    private ArrayList<Integer> listTags;

    @SerializedName("list_due")
    @Expose
    @ColumnInfo(name = "list_due")
    private long listDue;

    @SerializedName("list_notified")
    @Expose
    @ColumnInfo(name = "list_notified")
    private boolean notified;

    @SerializedName("project_id")
    @Expose
    @ColumnInfo(name = "project_id")
    private int projectId;

    @SerializedName("archived")
    @Expose
    @ColumnInfo(name = "archived")
    private boolean archived;

    @Ignore
    private boolean editing = false;

    @Ignore
    public Tasks(String listName, ArrayList<String> listItems, ArrayList<Boolean> listItemsChecked, ArrayList<Integer> listTags, DateTime time, boolean notified) {
        this.listName = listName;
        this.listItems = listItems;
        this.listItemsChecked = listItemsChecked;
        this.listTags = listTags;
        this.listDue = time.getMillis();
        this.notified = notified;
        this.listKey = ProjectsUtils.keyGenerator();
        this.archived = false;
        this.projectId = Current.project().getKey();
    }

    @Ignore
    public Tasks(String listName, ArrayList<String> listItems, ArrayList<Boolean> listItemsChecked, ArrayList<Integer> listTags, DateTime time, boolean notified, int listKey, int projectId, boolean archived) {
        this.listName = listName;
        this.listItems = listItems;
        this.listItemsChecked = listItemsChecked;
        this.listTags = listTags;
        this.listDue = time.getMillis();
        this.notified = notified;
        this.listKey = listKey;
        this.projectId = projectId;
        this.archived = archived;
    }

    public Tasks(int listKey, String listName, ArrayList<String> listItems, ArrayList<Boolean> listItemsChecked, ArrayList<Integer> listTags, long listDue, boolean notified, int projectId, boolean archived) {
        this.listKey = listKey;
        this.listName = listName;
        this.listItems = listItems;
        this.listItemsChecked = listItemsChecked;
        this.listTags = listTags;
        this.listDue = listDue;
        this.notified = notified;
        this.projectId = projectId;
        this.archived = archived;
    }


    public ArrayList<String> getListItems() {
        return listItems;
    }

    public void setListItems(ArrayList<String> listItems) {
        this.listItems = listItems;
    }

    public ArrayList<Boolean> getListItemsChecked() {
        return listItemsChecked;
    }

    public void setListItemsChecked(ArrayList<Boolean> listItemsChecked) {
        this.listItemsChecked = listItemsChecked;
    }

    public ArrayList<Integer> getListTags() {
        return listTags;
    }

    public long getListDue() {
        return listDue;
    }

    public void setListDue(long listDue) {
        this.listDue = listDue;
    }

    public String getListName() {
        return listName;
    }

    public void setListName(String listName) {
        this.listName = listName;
    }

    public boolean doesListContainTag(int tags) {
        if (!listTags.isEmpty()) {
            for (int i = 0; i < listTags.size(); i++) {
                if (listTags.get(i) == tags) {
                    return true;
                }

            }
        }
        return false;

    }

    public String getListItems(int pos) {
        return listItems.get(pos);
    }

    public ArrayList<String> getAllListItems() {
        return listItems;
    }

    public int getListItemsSize() {
        return listItems.size();
    }

    public boolean isListItems() {
        return !listItems.isEmpty();
    }

    public void setListItems(String item) {
        this.listItems.add(item);
    }

    public boolean getListItemsChecked(int pos) {

        if (listItemsChecked == null) {
            listItemsChecked = new ArrayList<>();
            for (int i = 1; i <= listItems.size(); i++) {
                listItemsChecked.add(false);
            }
        } else if (listItemsChecked.isEmpty() | listItemsChecked.size() == 0) {
            listItemsChecked = new ArrayList<>();
            for (int i = 1; i <= listItems.size(); i++) {
                listItemsChecked.add(false);
            }
        }
        return listItemsChecked.get(pos);
    }

    public void editListItemsChecked(boolean checked, int pos) {
        this.listItemsChecked.set(pos, checked);
    }

    public ArrayList<Boolean> getAllListItemsChecked() {
        return listItemsChecked;
    }

    public int getListTags(int pos) {
        Log.d("position", Integer.toString(pos));
        Log.d("position", Integer.toString(listTags.size()));
        return listTags.get(pos);
    }

    public void setListTags(int listPos, int tagPos) {
        this.listTags.set(listPos, tagPos);
    }

    public ArrayList<Integer> getAllListTags() {
        return listTags;
    }

    public int getListTagsSize() {
        return listTags.size();
    }

    public boolean isListTags() {
        return !listTags.isEmpty();
    }

    public void setListTags(ArrayList<Integer> listTags) {
        this.listTags = listTags;
    }

    public DateTime getDateTime() {
        return new DateTime(listDue);
    }

    public void setDateTime(DateTime datetime) {
        this.listDue = datetime.getMillis();
    }

    public boolean hasTime() {
        return getDateTime().getSecondOfMinute()==0;
    }

    public boolean hasDate() {
        if(getDateTime().equals(new DateTime(3000, 1, 1, 0, 0))){//fixes migration of old dates
            setDateTime(new DateTime(3000, 1, 1, 0, 59, 30));
        }
        return getDateTime().getSecondOfMinute()!=30;
    }


    public boolean isNotified() {
        return notified;
    }

    public void setNotified(boolean notified) {
        this.notified = notified;
    }

    public int getListKey() {
        return this.listKey;
    }

    public void setListKey(int listKey) {
        this.listKey = listKey;
    }

    public int getProjectId() {
        return projectId;
    }

    public void setProjectId(int projectId) {
        this.projectId = projectId;
    }

    public boolean isArchived() {
        return archived;
    }

    public void setArchived(boolean archived) {
        this.archived = archived;
    }

    public void normalizeAfterImport() {
        if (listTags == null) {
            listTags = new ArrayList<>();

        }
        if (listItems == null) {
            listItems = new ArrayList<>();

        }
        if (listName == null) {
            listName = "";
        }
    }

    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder();
        builder.append(getListName());
        builder.append(", " + getDateTime().toString("MMMM D, h:mm") + "\n");
        for (String s : getListItems()) {
            builder.append("- " + s + "\n");
        }
        builder.append("Tags: \n");
        for (int i : getListTags()) {
            builder.append("- Tag " + i + "\n");
        }
        return builder.toString();
    }

    public int compareTo(Tasks o, int filterMode) {
        Log.d("compareTasks", "filterMode: " + filterMode);
        if (filterMode == TIME_SORT) {
            Log.d("compareTasks", Integer.toString(compareTimes(o)));
            if (compareTimes(o) != 0) {
                return compareTimes(o);
            } else if(compareEditing(o)!=0){
                return compareEditing(o);
            }else if (compareAlphabetical(o) != 0) {
                return compareAlphabetical(o);
            } else {
                return compareLists(o);
            }

        } else {//Alphabetical sort
            if(compareEditing(o)!=0){
                return compareEditing(o);
            }else if (compareAlphabetical(o) != 0) {
                return compareAlphabetical(o);
            } else if (compareTimes(o) != 0) {
                return compareTimes(o);
            } else {
                return compareLists(o);
            }
        }
    }

    public int compareTimes(Tasks o) {
        return this.getDateTime().compareTo(o.getDateTime());
    }

    public int compareEditing(Tasks o) {
        if(this.isEditing()!=o.isEditing()){
            if(this.isEditing()){
                return 1;
            }else {
                return -1;
            }
        }else {
            return 0;
        }


    }

    public int compareAlphabetical(Tasks o) {
        return this.listName.compareTo(o.listName);
    }

    public int compareLists(Tasks o){
        return Integer.compare(this.getListItemsSize(), o.getListItemsSize());
    }

    public boolean isEditing() {
        return editing;
    }

    public void setEditing(boolean editing) {
        this.editing = editing;
    }


}

File: app/src/main/java/com/andb/apps/todo/objects/Tasks.kt
package com.andb.apps.todo.objects

import androidx.room.*
import com.andb.apps.todo.objects.reminders.LocationFence
import com.andb.apps.todo.objects.reminders.SimpleReminder
import com.andb.apps.todo.typeconverters.*
import com.andb.apps.todo.utilities.Current
import com.andb.apps.todo.utilities.ProjectsUtils
import com.andb.apps.todo.utilities.Values.SORT_TIME
import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName
import org.joda.time.DateTime
import java.util.ArrayList
import kotlin.Boolean
import kotlin.Cloneable
import kotlin.Comparator
import kotlin.Int
import kotlin.NoSuchElementException
import kotlin.String

@Entity(foreignKeys = [ForeignKey(entity = Project::class, parentColumns = ["key"], childColumns = ["project_id"], onDelete = ForeignKey.CASCADE)])
class Tasks : Cloneable {
    @SerializedName("key")
    @Expose
    @PrimaryKey
    val listKey: Int

    @SerializedName("list_name")
    @Expose
    @ColumnInfo(name = "list_name")
    var listName: String

    @SerializedName("list_items")
    @Expose
    @TypeConverters(ItemsConverter::class)
    @ColumnInfo(name = "list_items")
    val listItems: ArrayList<String>

    @SerializedName("list_items_checked")
    @Expose
    @TypeConverters(CheckedConverter::class)
    @ColumnInfo(name = "list_items_checked")
    val listItemsChecked: ArrayList<Boolean>

    @SerializedName("list_tags")
    @Expose
    @TypeConverters(TagConverter::class)
    @ColumnInfo(name = "list_tags")
    val listTags: ArrayList<Int>

    @SerializedName("list_times")
    @Expose
    @TypeConverters(SimpleReminderConverter::class)
    @ColumnInfo(name = "list_times")
    val timeReminders: ArrayList<SimpleReminder>

    @SerializedName("list_locations")
    @Expose
    @TypeConverters(LocationFenceConverter::class)
    @ColumnInfo(name = "list_locations")
    val locationReminders: ArrayList<LocationFence>

    @SerializedName("project_id")
    @Expose
    @ColumnInfo(name = "project_id")
    var projectId: Int

    @SerializedName("archived")
    @Expose
    @ColumnInfo(name = "archived")
    var isArchived: Boolean

    @Ignore
    var isEditing = false


    val isListItems: Boolean
        get() = !listItems.isEmpty()

    val listTagsSize: Int
        get() = listTags.size

    val isListTags: Boolean
        get() = !listTags.isEmpty()


    @Ignore
    @JvmOverloads
    constructor(listName: String, listItems: ArrayList<String>, listItemsChecked: ArrayList<Boolean>, listTags: ArrayList<Int>, timeReminders: ArrayList<SimpleReminder>,
                locationReminders: ArrayList<LocationFence>, projectId: Int = Current.projectKey(), archived: Boolean = false) : this(listName, listItems, listItemsChecked, listTags, timeReminders, locationReminders, ProjectsUtils.keyGenerator(), projectId, archived)

    constructor(listName: String, listItems: ArrayList<String>, listItemsChecked: ArrayList<Boolean>, listTags: ArrayList<Int>, timeReminders: ArrayList<SimpleReminder>,
                locationReminders: ArrayList<LocationFence>, listKey: Int, projectId: Int, archived: Boolean) {
        this.listKey = listKey
        this.listName = listName
        this.listItems = listItems
        this.listItemsChecked = listItemsChecked
        this.listTags = listTags
        this.timeReminders = timeReminders
        this.locationReminders = locationReminders
        this.projectId = projectId
        this.isArchived = archived
    }

    fun editListItemsChecked(checked: Boolean, pos: Int) {
        this.listItemsChecked[pos] = checked
    }


    override fun toString(): String {
        val builder = StringBuilder()
        builder.append(listName)
        builder.append(", " + nextReminderTime().toString("MMMM d, h:mm") + "\n")
        for (s in listItems) {
            builder.append("- $s\n")
        }
        if(listTags.isNotEmpty()) {
            builder.append("Tags: \n")
        }
        for (i in listTags) {
            builder.append("- Tag $i\n")
        }
        return builder.toString()
    }

    fun nextReminderTime(): DateTime {
        return nextReminder()?.asDateTime() ?: DateTime(3000, 1, 1, 0, 0, 59)
    }

    fun nextReminder(): SimpleReminder? {
        val sorted = timeReminders.sortedWith(Comparator { o1, o2 ->
            o1.asDateTime().compareTo(o2.asDateTime())
        })
        return try {
            sorted.first { !it.notified }
        } catch (e: NoSuchElementException) {
            if (!timeReminders.isEmpty()) {
                sorted[0]
            } else {
                null
            }
        }

    }

    fun hasLocationOrTrigger(locationKey: String): Boolean{
        val reminders: Boolean = locationReminders.map { it.key }.contains(locationKey.toInt())
        val triggers: Boolean = locationReminders.map { it.trigger?.key ?: -1 }.contains(locationKey.toInt())
        return reminders || triggers
    }

    fun findLocation(locationKey: String, isTrigger: Boolean = isTrigger(locationKey)): LocationFence{
        return if(isTrigger){
            locationReminders.first { it.trigger?.key.toString()==locationKey }
        }else{
            locationReminders.first { it.key.toString()==locationKey }
        }
    }

    fun isTrigger(locationKey: String): Boolean{
        return locationReminders.map { it.trigger?.key ?: -1 }.contains(locationKey.toInt())
    }

    fun compareTo(o: Tasks, filterMode: Int): Int {
        if (filterMode == SORT_TIME) {
            return if (compareTimes(o) != 0) {
                compareTimes(o)
            } else if (compareEditing(o) != 0) {
                compareEditing(o)
            } else if (compareAlphabetical(o) != 0) {
                compareAlphabetical(o)
            } else {
                compareLists(o)
            }

        } else {//Alphabetical sort
            return if (compareEditing(o) != 0) {
                compareEditing(o)
            } else if (compareAlphabetical(o) != 0) {
                compareAlphabetical(o)
            } else if (compareTimes(o) != 0) {
                compareTimes(o)
            } else {
                compareLists(o)
            }
        }
    }


    private fun compareTimes(o: Tasks): Int {
        return this.nextReminderTime().compareTo(o.nextReminderTime())
    }

    private fun compareEditing(o: Tasks): Int {
        return if (this.isEditing != o.isEditing) {
            if (this.isEditing) {
                1
            } else {
                -1
            }
        } else {
            0
        }


    }

    private fun compareAlphabetical(o: Tasks): Int {
        return this.listName.compareTo(o.listName)
    }

    private fun compareLists(o: Tasks): Int {
        return Integer.compare(this.listItems.size, o.listItems.size)
    }

    fun withKey(key: Int): Tasks {
        return Tasks(
            listName, listItems, listItemsChecked, listTags, timeReminders
                ?: ArrayList(), locationReminders ?: ArrayList(), key, projectId, isArchived
        )
    }


}
